exercises.py

```diff
@@ -1,28 +1,53 @@
-"""
-Recursion Problem Set: Coding Exercises
-"""
+"""Recursion Problem Set: Coding Exercises"""
 
 def reverse(s: str) -> str:
-    """
-    Reverse string using recursive approach
-    """
 
+    if len(s) == 0 or len(s) == 1:
+        return s
+    else:
+        return reverse(s[1:]) + s[0]    
+        
 def fib_rec(n: int) -> int:
-    """
-    Recursive fibonacci
-    """
+    """Recursive fibonacci"""
+    if n == 1 or n == 0:
+        return 1
+    
+    return fib_rec(n - 1) + fib_rec(n-2)
 
 def fib_dyn(n: int) -> int:
-    """
-    Fibonacci using dynamic programming (w/ a cache)
-    """
+    """Fibonacci using dynamic programming (w/ a cache)"""
+    cache = {}
```

Comment:
You actually want the cache outside of the function that is being recursively called :+1: